tests/csapi/sync_test.go

```diff
@@ -0,0 +1,116 @@
+package csapi_tests
+
+import (
+	"net/url"
+	"testing"
+	"time"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+)
+
+// Observes "first bug" from https://github.com/matrix-org/dendrite/pull/1394#issuecomment-687056673
+func TestCumulativeJoinLeaveJoinSync(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintOneToOneRoom)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+	bob := deployment.Client(t, "hs1", "@bob:hs1")
+
+	roomID := bob.CreateRoom(t, map[string]interface{}{
+		"preset": "public_chat",
+	})
+
+	var since string
+
+	// Get floating next_batch from before joining at all
+	res := alice.MustDoFunc(t, "GET", []string{"_matrix", "client", "r0", "sync"}, client.WithQueries(url.Values{
+		"timeout": []string{"0"},
+	}))
+	body := client.ParseJSON(t, res)
+	since = client.GetJSONFieldStr(t, body, "next_batch")
+
+	alice.JoinRoom(t, roomID, nil)
+
+	// We can't SyncUntil or else "pollute" the sync internal state, we must do as-if we're a slow client.
+	// Thus, two-second delays.
+	// WARNING: If you're wondering why this test is failing in a flaky fashion, and are looking at this test,
+	// this may the reason why, sorry :(
+	time.Sleep(2 * time.Second)
```

Comment:
The bug report makes no mention of syncing between the join/leaves, so we shouldn't need to sleep here. It should be fine to join->leave->join -> sleep(1 second) -> do the sync. Try not to sleep a lot as this will slow down complement runs for everyone all the time.

Reply:
Just because the bug report makes no mention of syncing, doesnt mean that *doing* the sync is allowed.

I'm taking a conservative approach here by sleeping here, but i'll give this a look again later.